service/history/queues/priority_assigner.go

```diff
@@ -25,77 +25,66 @@
 package queues
 
 import (
-	"sync"
+	"fmt"
 
 	"go.temporal.io/api/serviceerror"
 	enumsspb "go.temporal.io/server/api/enums/v1"
-	"go.temporal.io/server/common/dynamicconfig"
-	"go.temporal.io/server/common/metrics"
 	"go.temporal.io/server/common/namespace"
-	"go.temporal.io/server/common/quotas"
 	"go.temporal.io/server/common/tasks"
 )
 
 type (
 	// PriorityAssigner assigns priority to task executables
 	PriorityAssigner interface {
-		Assign(Executable) error
+		Assign(Executable) tasks.Priority
 	}
 
 	PriorityAssignerOptions struct {
-		CriticalRetryAttempts dynamicconfig.IntPropertyFn
+		CriticalRetryAttempts int
 	}
 
 	priorityAssignerImpl struct {
 		currentClusterName string
 		namespaceRegistry  namespace.Registry
-		metricsProvider    metrics.MetricsHandler
 		options            PriorityAssignerOptions
-
-		sync.RWMutex
-		rateLimiters map[string]quotas.RateLimiter
 	}
+
+	// noopPriorityAssigner always assign high priority to tasks
+	// it should only be used in tests
+	noopPriorityAssigner struct{}
 )
 
 func NewPriorityAssigner(
 	currentClusterName string,
 	namespaceRegistry namespace.Registry,
 	options PriorityAssignerOptions,
-	metricsProvider metrics.MetricsHandler,
 ) PriorityAssigner {
 	return &priorityAssignerImpl{
 		currentClusterName: currentClusterName,
 		namespaceRegistry:  namespaceRegistry,
-		metricsProvider:    metricsProvider.WithTags(metrics.OperationTag(OperationTaskPriorityAssigner)),
 		options:            options,
-		rateLimiters:       make(map[string]quotas.RateLimiter),
 	}
 }
 
-func (a *priorityAssignerImpl) Assign(executable Executable) error {
+func (a *priorityAssignerImpl) Assign(executable Executable) tasks.Priority {
```

Comment:
Because having a `SetPriority()` method in Executable interface which is only invoked from inside the Executable feels very weird to me.

I don't feel strongly about the naming, calculating & returning the priority still feels like assign to me.

Reply:
Existing code call SetPriority() from outside of executable. Why we move away from that model?

Reply:
Please check the PR description. 
Existing code do priority assigning upon task submission, which may change the task channel it goes to, making us unable to determine which task channel is full and which is not. 

So instead of assigning priority upon submit, I choose to let the executable manage the priority itself, and make sure it's updated before adding itself to the rescheduler.

Reply:
Some more context on this. Strictly speaking, the assumption of this change is task priority won't change during reschedule backoff, which is not true as namespace can failover during the backoff period. 

For existing failover implementation, this won't be an issue because upon failover, a new failover queue will be created and reload all pending tasks. Those newly loaded tasks will have the updated priority. Existing tasks will be no-op tasks, so we don't care about their priority/latency etc.

For multi-cursor implementation, the fact that task priority can change affects not only tasks in rescheduler but also tasks buffered in the host level task scheduler (worker pool). And for the host level task scheduler, there's no way to change the task channel after submitting it. As long as we can solve the problem for the task scheduler, the issue for rescheduler can be solved as well.

So "doing priority assignment before adding to rescheduler" won't have any negative impact for existing xdc logic. Nor will it make the problem for multi-cursor more complicated. It also allows task rescheduler to reason about when it should stop submitting tasks.

Let's sync offline on how to handle namespace failover in multi-cursor implementation.